File: frontend/src/tokens/multiToken.js
/**
 * AiFund Pay — Multi-Token Payment Support
 * Accept USDT, USDC, ETH, BTC, SOL and any ERC-20/SPL token.
 * https://aifund.com/pay
 *
 * Adds to config.js:
 *   - Multiple token definitions per chain
 *   - Token selection in PaymentFlow
 *   - Price conversion via CoinGecko API
 */

// Default supported tokens with contract addresses
export const SUPPORTED_TOKENS = {
  // Stablecoins (1:1 USD, no conversion needed)
  USDT: {
    symbol: 'USDT',
    name: 'Tether USD',
    decimals: 6,
    isStablecoin: true,
    icon: '💵',
    chains: {
      trc20: { contract: 'TR7NHqjeKQxGTCi8q8ZY4pL8otSzgjLj6t' },
      erc20: { contract: '0xdac17f958d2ee523a2206206994597c13d831ec7' },
      bsc: { contract: '0x55d398326f99059ff775485246999027b3197955' },
      arb: { contract: '0xfd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9' },
      polygon: { contract: '0xc2132D05D31c914a87C6611C10748AaCBb0B0cE' },
      sol: { contract: 'Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB' },
    },
  },
  USDC: {
    symbol: 'USDC',
    name: 'USD Coin',
    decimals: 6,
    isStablecoin: true,
    icon: '🔵',
    chains: {
      erc20: { contract: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48' },
      arb: { contract: '0xaf88d065e77c8cC2239327C5EDb3A432268e5831' },
      polygon: { contract: '0x3c499c542cef5e3811e1192ce70d8cc03d5c3359' },
      base: { contract: '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913' },
      sol: { contract: 'EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v' },
    },
  },
  'USDC.e': {
    symbol: 'USDC.e',
    name: 'Bridged USDC (Polygon)',
    decimals: 6,
    isStablecoin: true,
    icon: '🔵',
    chains: {
      polygon: { contract: '0x2791bca1f2de4661ed88a30c99a7a9449aa84174' },
    },
  },

  // Native tokens (need price conversion)
  ETH: {
    symbol: 'ETH',
    name: 'Ethereum',
    decimals: 18,
    isStablecoin: false,
    icon: '🔷',
    coingeckoId: 'ethereum',
    chains: {
      erc20: { native: true },
      arb: { native: true },
      base: { native: true },
    },
  },
  POL: {
    symbol: 'POL',
    name: 'Polygon',
    decimals: 18,
    isStablecoin: false,
    icon: '🟪',
    coingeckoId: 'matic-network',
    chains: {
      polygon: { native: true },
    },
  },
  BTC: {
    symbol: 'BTC',
    name: 'Bitcoin',
    decimals: 8,
    isStablecoin: false,
    icon: '₿',
    coingeckoId: 'bitcoin',
    chains: {
      bitcoin: { native: true },
    },
  },
  SOL: {
    symbol: 'SOL',
    name: 'Solana',
    decimals: 9,
    isStablecoin: false,
    icon: '🟣',
    coingeckoId: 'solana',
    chains: {
      sol: { native: true },
    },
  },
  BNB: {
    symbol: 'BNB',
    name: 'BNB',
    decimals: 18,
    isStablecoin: false,
    icon: '🟡',
    coingeckoId: 'binancecoin',
    chains: {
      bsc: { native: true },
    },
  },
};

/**
 * Get current USD price for a token via CoinGecko
 * @param {string} coingeckoId - e.g. 'ethereum', 'bitcoin', 'solana'
 * @returns {number} USD price
 */
export async function getTokenPrice(coingeckoId) {
  try {
    const res = await fetch(
      `https://api.coingecko.com/api/v3/simple/price?ids=${coingeckoId}&vs_currencies=usd`
    );
    const data = await res.json();
    return data[coingeckoId]?.usd || 0;
  } catch (e) {
    console.error('Price fetch error:', e);
    return 0;
  }
}

/**
 * Calculate how much of a token is needed for a USD amount
 * @param {string} tokenSymbol - e.g. 'ETH', 'BTC', 'SOL'
 * @param {number} usdAmount - Amount in USD
 * @returns {Object} { amount, price, symbol }
 */
export async function calculateTokenAmount(tokenSymbol, usdAmount) {
  const token = SUPPORTED_TOKENS[tokenSymbol];
  if (!token) return null;

  if (token.isStablecoin) {
    return { amount: usdAmount, price: 1, symbol: tokenSymbol };
  }

  const price = await getTokenPrice(token.coingeckoId);
  if (!price) return null;

  const amount = usdAmount / price;
  // Round to reasonable decimals
  const decimals = token.symbol === 'BTC' ? 8 : token.symbol === 'ETH' ? 6 : 4;
  return {
    amount: parseFloat(amount.toFixed(decimals)),
    price,
    symbol: tokenSymbol,
  };
}

/**
 * Get available tokens for a specific chain
 * @param {string} chainId - e.g. 'erc20', 'sol', 'bsc'
 * @returns {Array} List of tokens available on this chain
 */
export function getTokensForChain(chainId) {
  const available = [];
  for (const [symbol, token] of Object.entries(SUPPORTED_TOKENS)) {
    if (token.chains[chainId]) {
      available.push({
        symbol,
        name: token.name,
        icon: token.icon,
        isStablecoin: token.isStablecoin,
        isNative: token.chains[chainId].native || false,
      });
    }
  }
  return available;
}

File: backend/multi_token.py
"""
AiFund Pay — Multi-Token Payment Support
Accept USDT, USDC, ETH, BTC, SOL and custom ERC-20/SPL tokens.
https://aifund.com/pay

Adds to payment_verify.py:
  - Token-aware verification (check specific contract or native transfer)
  - Price conversion via CoinGecko
  - Multi-token balance tracking
"""
from fastapi import APIRouter
from typing import Dict, Optional
import httpx
import logging

logger = logging.getLogger(__name__)

# Token contract addresses for verification
TOKEN_CONTRACTS = {
    "USDT": {
        "erc20": "0xdac17f958d2ee523a2206206994597c13d831ec7",
        "bsc": "0x55d398326f99059ff775485246999027b3197955",
        "arb": "0xfd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
        "trc20": "TR7NHqjeKQxGTCi8q8ZY4pL8otSzgjLj6t",
    },
    "USDC": {
        "erc20": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
        "arb": "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
        "base": "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
    },
    "ETH": {"erc20": "native", "arb": "native", "base": "native"},
    "BNB": {"bsc": "native"},
    "SOL": {"sol": "native"},
}

COINGECKO_IDS = {
    "ETH": "ethereum",
    "BTC": "bitcoin",
    "SOL": "solana",
    "BNB": "binancecoin",
}


async def get_usd_price(token_symbol: str) -> float:
    """Get current USD price for a non-stablecoin token."""
    if token_symbol in ("USDT", "USDC"):
        return 1.0
    
    cg_id = COINGECKO_IDS.get(token_symbol)
    if not cg_id:
        return 0
    
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.get(
                f"https://api.coingecko.com/api/v3/simple/price?ids={cg_id}&vs_currencies=usd"
            )
            if resp.status_code == 200:
                data = resp.json()
                return data.get(cg_id, {}).get("usd", 0)
    except Exception as e:
        logger.error(f"Price fetch error for {token_symbol}: {e}")
    
    # Fallback prices
    fallbacks = {"ETH": 3500, "BTC": 100000, "SOL": 180, "BNB": 580}
    return fallbacks.get(token_symbol, 0)


def calculate_token_amount(usd_amount: float, token_symbol: str, price: float) -> float:
    """Calculate how much token is needed for a USD amount."""
    if token_symbol in ("USDT", "USDC"):
        return usd_amount
    if price <= 0:
        return 0
    return round(usd_amount / price, 8)


def create_multi_token_router(db) -> APIRouter:
    """Create router for multi-token price and conversion."""
    router = APIRouter()

    @router.get("/tokens/prices")
    async def get_token_prices():
        """Get current USD prices for all supported tokens."""
        prices = {}
        for symbol in ["USDT", "USDC", "ETH", "BTC", "SOL", "BNB"]:
            prices[symbol] = await get_usd_price(symbol)
        return {"prices": prices}

    @router.get("/tokens/convert")
    async def convert_amount(usd_amount: float, token: str = "ETH"):
        """Convert a USD amount to token amount."""
        price = await get_usd_price(token)
        amount = calculate_token_amount(usd_amount, token, price)
        return {
            "usd_amount": usd_amount,
            "token": token,
            "token_amount": amount,
            "price": price,
        }

    @router.get("/tokens/supported")
    async def get_supported_tokens():
        """List all supported tokens and their chains."""
        tokens = {}
        for symbol, chains in TOKEN_CONTRACTS.items():
            tokens[symbol] = {
                "symbol": symbol,
                "chains": list(chains.keys()),
                "is_stablecoin": symbol in ("USDT", "USDC"),
            }
        return {"tokens": tokens}

    return router
